azurerm_marketplace_agreement in conjunction with azurerm_linux_virtual_machine causes AzureRM to crash

<!--- Please keep this note for the community --->

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

<!--- Thank you for keeping this note for the community --->

### Terraform (and AzureRM Provider) Version
Terraform Version 1.0.4
AzureRM Version 2.71.0
<!--- Please run `terraform -v` to show the Terraform core version and provider version(s). If you are not running the latest version of Terraform or the provider, please upgrade because your issue may have already been fixed. [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions). --->

### Affected Resource(s)

azurerm_marketplace_agreement
azurerm_linux_virtual_machine

### Terraform Configuration Files

<!--- Information about code formatting: https://help.github.com/articles/basic-writing-and-formatting-syntax/#quoting-code --->

```hcl
resource "azurerm_marketplace_agreement" "barracuda" {
  publisher = "barracudanetworks"
  offer     = "waf"
  plan      = "hourly"
}
```
Using a marketplace image storage_image_reference in azurerm_linux_virtual_machine

### Debug Output

<!---
Please provide a link to a GitHub Gist containing the complete debug output. Please do NOT paste the debug output in the issue; just paste a link to the Gist.

To obtain the debug output, see the [Terraform documentation on debugging](https://www.terraform.io/docs/internals/debugging.html).
--->

### Panic Output
Error: Plugin did not respond
│ 
│   with azurerm_marketplace_agreement.app,
│   on TOSAgreement.tf line 1, in resource "azurerm_marketplace_agreement" "app":
│    1: resource "azurerm_marketplace_agreement" "app" {
│ 
│ The plugin encountered an error, and failed to respond to the
│ plugin.(*GRPCProvider).ApplyResourceChange call. The plugin logs may
│ contain more details.

Stack trace from the terraform-provider-azurerm_v2.71.0_x5 plugin:
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x48e31ee]
goroutine 168 [running]:
github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/compute.resourceMarketplaceAgreementCreateUpdate(0xc001437780, 0x57e1000, 0xc000a48e00, 0x0, 0x0)
	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/compute/marketplace_agreement_resource.go:99 +0x12ce
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).create(0xc0008fbdc0, 0x6787328, 0xc001771300, 0xc001437780, 0x57e1000, 0xc000a48e00, 0x0, 0x0, 0x0)
	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-azurerm/vendor/github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema/resource.go:318 +0x1ee
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).Apply(0xc0008fbdc0, 0x6787328, 0xc001771300, 0xc000160310, 0xc001a70e80, 0x57e1000, 0xc000a48e00, 0x0, 0x0, 0x0, ...)
	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-azurerm/vendor/github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema/resource.go:456 +0x67b
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*GRPCProviderServer).ApplyResourceChange(0xc000317d10, 0x6787328, 0xc001771300, 0xc001441b80, 0xc001771300, 0x5cc9180, 0xc000a45800)
	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-azurerm/vendor/github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema/grpc_provider.go:955 +0x8ef
github.com/hashicorp/terraform-plugin-go/tfprotov5/server.(*server).ApplyResourceChange(0xc001b3d7a0, 0x67873d0, 0xc001771300, 0xc000993f10, 0xc001b3d7a0, 0xc000a458c0, 0xc001cadba0)
	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-azurerm/vendor/github.com/hashicorp/terraform-plugin-go/tfprotov5/server/server.go:332 +0xb5
github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/tfplugin5._Provider_ApplyResourceChange_Handler(0x5cc9180, 0xc001b3d7a0, 0x67873d0, 0xc000a458c0, 0xc00174f1a0, 0x0, 0x67873d0, 0xc000a458c0, 0xc000c90000, 0x1dc)
	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-azurerm/vendor/github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/tfplugin5/tfplugin5_grpc.pb.go:380 +0x214
google.golang.org/grpc.(*Server).processUnaryRPC(0xc0002388c0, 0x67d3ab8, 0xc00025a780, 0xc001686d80, 0xc000676ae0, 0x9d7cfa0, 0x0, 0x0, 0x0)
	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-azurerm/vendor/google.golang.org/grpc/server.go:1292 +0x52b
google.golang.org/grpc.(*Server).handleStream(0xc0002388c0, 0x67d3ab8, 0xc00025a780, 0xc001686d80, 0x0)
	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-azurerm/vendor/google.golang.org/grpc/server.go:1617 +0xd0c
google.golang.org/grpc.(*Server).serveStreams.func1.2(0xc000137cd0, 0xc0002388c0, 0x67d3ab8, 0xc00025a780, 0xc001686d80)
	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-azurerm/vendor/google.golang.org/grpc/server.go:940 +0xab
created by google.golang.org/grpc.(*Server).serveStreams.func1
	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-azurerm/vendor/google.golang.org/grpc/server.go:938 +0x1fd
Error: The terraform-provider-azurerm_v2.71.0_x5 plugin crashed!
This is always indicative of a bug within the plugin. It would be immensely
helpful if you could report the crash with the plugin's maintainers so that it
can be fixed. The output above should help diagnose the issue.
<!--- If Terraform produced a panic, please provide a link to a GitHub Gist containing the output of the `crash.log`. --->

### Expected Behaviour

<!--- What should have happened? --->

### Actual Behaviour

<!--- What actually happened? --->

### Steps to Reproduce

<!--- Please list the steps required to reproduce the issue. --->

1. `terraform apply`

### Important Factoids

<!--- Are there anything atypical about your accounts that we should know? For example: Running in a Azure China/Germany/Government? --->
None

### References

<!---
Information about referencing Github Issues: https://help.github.com/articles/basic-writing-and-formatting-syntax/#referencing-issues-and-pull-requests

Are there any other GitHub issues (open or closed) or pull requests that should be linked here? Such as vendor documentation?
--->

* #0000

--- a/internal/services/compute/marketplace_agreement_resource.go
+++ b/internal/services/compute/marketplace_agreement_resource.go
@@ -69,18 +69,17 @@ func resourceMarketplaceAgreement() *pluginsdk.Resource {
 func resourceMarketplaceAgreementCreateUpdate(d *pluginsdk.ResourceData, meta interface{}) error {
 	client := meta.(*clients.Client).Compute.MarketplaceAgreementsClient
 	ctx, cancel := timeouts.ForCreateUpdate(meta.(*clients.Client).StopContext, d)
+	subscriptionId := meta.(*clients.Client).Account.SubscriptionId
 	defer cancel()
 
-	offer := d.Get("offer").(string)
-	plan := d.Get("plan").(string)
-	publisher := d.Get("publisher").(string)
+	id := parse.NewPlanID(subscriptionId, d.Get("publisher").(string), d.Get("offer").(string), d.Get("plan").(string))
 
-	log.Printf("[DEBUG] Retrieving the Marketplace Terms for Publisher %q / Offer %q / Plan %q", publisher, offer, plan)
+	log.Printf("[DEBUG] retreiving %s", id)
 
-	term, err := client.Get(ctx, publisher, offer, plan)
+	term, err := client.Get(ctx, id.AgreementName, id.OfferName, id.Name)
 	if err != nil {
 		if !utils.ResponseWasNotFound(term.Response) {
-			return fmt.Errorf("retrieving the Marketplace Terms for Publisher %q / Offer %q / Plan %q: %s", publisher, offer, plan, err)
+			return fmt.Errorf("retrieving %s: %s", id, err)
 		}
 	}
 
@@ -92,37 +91,32 @@ func resourceMarketplaceAgreementCreateUpdate(d *pluginsdk.ResourceData, meta in
 	}
 
 	if accepted {
-		agreement, err := client.GetAgreement(ctx, publisher, offer, plan)
+		agreement, err := client.GetAgreement(ctx, id.AgreementName, id.OfferName, id.Name)
 		if err != nil {
 			if !utils.ResponseWasNotFound(agreement.Response) {
-				return fmt.Errorf("retrieving agreement for Publisher %q / Offer %q / Plan %q: %s", publisher, offer, plan, err)
+				return fmt.Errorf("retrieving %s: %s", id, err)
 			}
 		}
-		return tf.ImportAsExistsError("azurerm_marketplace_agreement", *agreement.ID)
+		return tf.ImportAsExistsError("azurerm_marketplace_agreement", id.ID())
 	}
 
-	terms, err := client.Get(ctx, publisher, offer, plan)
+	terms, err := client.Get(ctx, id.AgreementName, id.OfferName, id.Name)
 	if err != nil {
-		return fmt.Errorf("retrieving the Marketplace Terms for Publisher %q / Offer %q / Plan %q: %s", publisher, offer, plan, err)
+		return fmt.Errorf("retrieving %s: %s", id, err)
 	}
 	if terms.AgreementProperties == nil {
-		return fmt.Errorf("retrieving the Marketplace Terms for Publisher %q / Offer %q / Plan %q: AgreementProperties was nil", publisher, offer, plan)
+		return fmt.Errorf("retrieving %s: AgreementProperties was nil", id)
 	}
 
 	terms.AgreementProperties.Accepted = utils.Bool(true)
 
-	log.Printf("[DEBUG] Accepting the Marketplace Terms for Publisher %q / Offer %q / Plan %q", publisher, offer, plan)
-	if _, err := client.Create(ctx, publisher, offer, plan, terms); err != nil {
-		return fmt.Errorf("accepting Terms for Publisher %q / Offer %q / Plan %q: %s", publisher, offer, plan, err)
+	log.Printf("[DEBUG] Accepting the Marketplace Terms for %s", id)
+	if _, err := client.Create(ctx, id.AgreementName, id.OfferName, id.Name, terms); err != nil {
+		return fmt.Errorf("accepting Terms for %s: %s", id, err)
 	}
-	log.Printf("[DEBUG] Accepted the Marketplace Terms for Publisher %q / Offer %q / Plan %q", publisher, offer, plan)
+	log.Printf("[DEBUG] Accepted the Marketplace Terms for %s", id)
 
-	agreement, err := client.GetAgreement(ctx, publisher, offer, plan)
-	if err != nil {
-		return fmt.Errorf("retrieving agreement for Publisher %q / Offer %q / Plan %q: %s", publisher, offer, plan, err)
-	}
-
-	d.SetId(*agreement.ID)
+	d.SetId(id.ID())
 
 	return resourceMarketplaceAgreementRead(d, meta)
 }

--- a/internal/services/compute/marketplace_agreement_resource.go
+++ b/internal/services/compute/marketplace_agreement_resource.go
@@ -74,7 +74,7 @@ func resourceMarketplaceAgreementCreateUpdate(d *pluginsdk.ResourceData, meta in
 
 	id := parse.NewPlanID(subscriptionId, d.Get("publisher").(string), d.Get("offer").(string), d.Get("plan").(string))
 
-	log.Printf("[DEBUG] retreiving %s", id)
+	log.Printf("[DEBUG] retrieving %s", id)
 
 	term, err := client.Get(ctx, id.AgreementName, id.OfferName, id.Name)
 	if err != nil {

